build.sbt

```diff
@@ -49,6 +49,16 @@ scalacOptions ++= Seq(
   "-Xlint",
   "-encoding", "UTF-8"
 )
+ThisBuild / scalacOptions ++= {
+  CrossVersion.partialVersion(scalaVersion.value) match {
+    case Some((2, n)) if n == 12 || n == 13 =>
+      Seq("-opt-inline-from:<sources>",
+        "-opt:l:inline")
```

Comment:
In general I'm somewhat skeptical of inlining applied to sbt plugins, esp without benchmark showing it does anything beyond modern JVM JITing. The downside could be that the compilation takes longer and there might be unforseen effects like Zinc not working supposedly.

Reply:
@eed3si9n Do you still feel strong about this? I would prefer to either merge it or close it rather than just leave it sitting  here.

Also as a FYI incase its clear

> The downside could be that the compilation takes longer and there might be unforseen effects like Zinc not working supposedly.

This can't really happen, the inliner only effects bytecode that is generated by the sources of the project itself (i.e. everything in `com.typesafe.sbt.osgi` so it shouldn't have any effect on Zinc, thats what the `-opt-inline-from:<sources>` does (its the safest and most conservative option meant to be used for libraries). Also a large number of sbt community plugins (and sbt plugins in general) are using the inliner and there hasn't been any reported problems.

Reply:
https://docs.scala-lang.org/overviews/compiler-options/optimizer.html mentions its lack of incremental compilation support in multiple places:

> - Don’t enable the optimizer during development: it breaks incremental compilation, and it makes the compiler slower. Only enable it for testing, on CI, and to build releases.
>
> - Note that in an incremental compilation, inlining would only happen within the sources being re-compiled – but in any case, it is recommended to only enable the optimizer in CI and release builds (and to run clean before building).

Maybe the doc was written in a conservative way, an in practice nothing bad happens, but optimizers are usually written with only clean compilation in mind. There's a key called `insideCI`, which we could use to enable it when it's `true`?

https://github.com/sbt/sbt/blob/1.10.x/main/src/main/scala/sbt/Keys.scala#L580

Reply:
> Maybe the doc was written in a conservative way, an in practice nothing bad happens, but optimizers are usually written with only clean compilation in mind. 

Yeah from practice I haven't had a case where this is a problem when using `-opt-inline-from:<sources>`. Where you can typically hit problems is if you use global level inlining i.e. `-opt:inline:**` since for this to work properly, the entire set of Scala sources needs to be sent to scalac, hence the clean requirement.

> Maybe the doc was written in a conservative way, an in practice nothing bad happens, but optimizers are usually written with only clean compilation in mind. There's a key called `insideCI`, which we could use to enable it when it's `true`?
> 
> https://github.com/sbt/sbt/blob/1.10.x/main/src/main/scala/sbt/Keys.scala#L580

TIL, I had no idea of the existence of `insideCI` so thats something I can indeed use, will get back to this later.

Reply:
@eed3si9n I have updated PR to use `insideCI`